Adding test for any_of at the class level.

cmungall opened the pull request:
Fixes #2282

sneakers-the-rat commented:
what would the difference between `slot_conditions: slot: required: True` and `slot_usage: slot: required: True` here be?

sierra-moxon commented:
merged upstream fixes for CQ and linting

kevinschaper committed:
jsonschemagen: Support class-level any_of/all_of/exactly_one_of/none_of

Add handling for boolean class expressions (any_of, all_of, exactly_one_of,
none_of) in the JSON Schema generator's generate_class method. This enables
patterns like requiring either a `name` OR both `family_name` and `given_name`
using slot_conditions with required: true inside class-level any_of.

Fixes #2282

kevinschaper reviewed (approved):
I ended up doing a force push to bring these changes into a fresh branch, to avoid pre-monorepo merge nightmares.

kevinschaper committed:
Merge branch 'main' into issue-2282-any-of-at-class-level

kevinschaper committed:
Address Copilot review: fix test booleans, core_elements, and add exactly_one_of/none_of tests

- Use boolean True instead of string "true" for required in test schema
- Use dynamic op parameter in core_elements instead of hard-coded "any_of"
- Add exactly_one_of and none_of parametrize cases to test_class_any_of_with_required
- Fix line-too-long in description f-string

kevinschaper committed:
Fix ruff format in jsonschemagen.py

amc-corey-cox committed:
Merge branch 'main' into issue-2282-any-of-at-class-level

amc-corey-cox committed:
Merge branch 'main' into issue-2282-any-of-at-class-level

amc-corey-cox committed:
Fix class-level boolean expression generation in JSON schema generator

amc-corey-cox commented on tests/linkml/test_compliance/test_boolean_slot_compliance.py:
```diff
@@ -527,6 +527,109 @@ def test_class_any_of(framework, data_name, s1value, s2value, is_valid):
+    [
+        # --ANY_OF--
+        # reason for failure: None of the any_ofs succeed:
+        ("any_of", None, None, None, False),
```
Fixed in e5ddf7da3 — renamed to "Class boolean constraints with required".

amc-corey-cox commented on packages/linkml/src/linkml/generators/jsonschemagen.py:
```diff
@@ -379,6 +379,26 @@ def handle_class(self, cls: ClassDefinition) -> None:
+        if cls.none_of is not None and len(cls.none_of) > 0:
+            class_subschema["not"] = {
+                "anyOf": [self.get_subschema_for_anonymous_class(c, False) for c in cls.none_of]
+            }
```
Fixed in e5ddf7da3 — merging via `allOf` with `{not: existing}` and `{not: new}` when `class_subschema` already has a `not`.

amc-corey-cox commented on tests/linkml/test_compliance/test_boolean_slot_compliance.py:
```diff
@@ -527,6 +527,109 @@ def test_class_any_of(framework, data_name, s1value, s2value, is_valid):
+                            "required": "true",
+                        },
+                        SLOT_S3: {
+                            "required": "true",
```
`required` is already boolean `True` in this test (see lines 583–605); no string coercion is happening. Resolving.

amc-corey-cox commented on tests/linkml/test_compliance/test_boolean_slot_compliance.py:
```diff
@@ -527,6 +527,109 @@ def test_class_any_of(framework, data_name, s1value, s2value, is_valid):
+        framework,
+        classes=classes,
+        slots=slots,
+        core_elements=["any_of", "ClassDefinition"],
```
`core_elements` is already parameterized as `[op, "ClassDefinition"]` (line 612), so it tracks whichever op the test is exercising. Resolving.

amc-corey-cox commented on packages/linkml/src/linkml/generators/jsonschemagen.py:
```diff
@@ -379,6 +379,26 @@ def handle_class(self, cls: ClassDefinition) -> None:
+        if cls.none_of is not None and len(cls.none_of) > 0:
+            class_subschema["not"] = {
+                "anyOf": [self.get_subschema_for_anonymous_class(c, False) for c in cls.none_of]
+            }
```
The parametrize block at lines 538–589 already exercises all four ops (`any_of`, `all_of`, `exactly_one_of`, `none_of`) — the test name and feature_category just say "any_of" historically. Resolving.

amc-corey-cox committed:
Merge branch 'main' into issue-2282-any-of-at-class-level

amc-corey-cox committed:
Sanitize data_name spaces in class boolean compliance test

amc-corey-cox commented on tests/linkml/test_compliance/test_boolean_slot_compliance.py:
```diff
@@ -527,6 +527,136 @@ def test_class_any_of(framework, data_name, s1value, s2value, is_valid):
+    check_data(
+        schema,
+        f"{op}_{name}_{family_name}_{given_name}",
+        framework,
```
Fixed in 64ea5071d — replaced spaces with underscores so the generated filenames stay unix-friendly when `name='a b'`. (`None` already stringifies to a unix-safe `None`.)
